### PR TITLE
Compressor: add UADK support

### PR DESCRIPTION
**Pull request by Svelar**
This patch add UADK Acceleration for zlib compressor

[UADK: Userspace Acceleration Development Kit](https://github.com/Linaro/uadk)

1. Unity: one set of interfaces to support enc and comp acceleration
2. Efficiency: userspace zerocopy based on IOMMU & SVA(shared virtual address)
3. Security: the IOMMU limits the access rights and security boundaries of devices and processes
4. Compatibility: all acc devices can join this opensource project

Test on Kunpeng 920 platform, compression offload based on UadkAccel:

1. save almost 50% cpu usage compared with no-isal compression in RBD 4M workload
2. save almost 40% cpu usage compared with no-isal compression in RGW put op (4M) workload
3. lower cpu usage, better performance

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by kevinzs2048** [comment]
Some background info for UADK:
UADK is an open source accelerator framework, the kernel support part is [UACCE](https://docs.kernel.org/misc-devices/uacce.html), which has been merged in Kernel for several years, targeting to provide Shared Virtual Addressing (SVA) between accelerators and processes. UADK provider users with a unified programming interface to efficiently harness the capabilities of hardware accelerators. It furnishes users with fundamental library and driver support. Now UADK can offer the hardware accelerator engine(e.g: Kunpeng KAE), Arm SVE and Crypto Extension CPU. UADK is hosted by Linaro: 

**Expansibility**, e.g: other silicon vendor to leverage UADK:
- SVA is needed
- Integrate the encrpytion/compressor accelerator driver to UACCE framework
- No change in UADK user mode. Just integrate UADK in the workload.

UADK now has already supports different community for compressor and encryption:
image

**Review by badone — APPROVED** [review]
This looks good to me but let's get input from the rbd and rgw teams.

**Review by cbodley — COMMENTED** [review]
very cool 👍 

can this library support any of our compression algorithms other than zlib?

File: src/compressor/zlib/ZlibCompressor.cc
Comment (idryomov) [inline src/compressor/zlib/ZlibCompressor.cc:211]: Unlike ISAL and QAT, this falls through to other ways to compress on errors.  Why is that?  Can UADK be enabled but all of a sudden stop working?

(The same question goes for the two decomression methods.)

**Review by idryomov — COMMENTED** [review]
> [...] let's get input from the rbd and rgw teams.

No objection from RBD.  Commenting on some things that caught my eye.

**Comment by Svelar** [comment]
> very cool 👍
> 
> can this library support any of our compression algorithms other than zlib?

For compression, UADK only supports zlib, gzip and zstd(except decompression).

File: cmake/modules/Builduadk.cmake
Comment (Svelar) [reply cmake/modules/Builduadk.cmake:10]: That is also my opinion. But this library just fixed a build error recently and can't backport to released branch or tag. I would modify the `GIT_TAG` once next TAG released.

File: CMakeLists.txt
Comment (Svelar) [reply CMakeLists.txt:386]: Make sense, will do.

File: cmake/modules/Builduadk.cmake
Comment (Svelar) [reply cmake/modules/Builduadk.cmake:18]: Yes, as first edition, I just used UADK default build way. Will refactor this part.

File: cmake/modules/Builduadk.cmake
Comment (cbodley) [inline cmake/modules/Builduadk.cmake:8]: can this be a static library? that would avoid the extra effort to a) package and install the shared lib with ceph packages, and b) rename things to avoid conflicts with existing libraries installed on the system

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:281]: Yes, handle in `wd_comp_alloc_sess` of UADK.

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:379]: Same as above

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:377]: https://github.com/ceph/ceph/blob/main/src/compressor/Compressor.h#L89 declared here.

File: src/compressor/zlib/ZlibCompressor.cc
Comment (Svelar) [reply src/compressor/zlib/ZlibCompressor.cc:211]: For initial intention here is that if UADK failed working somehow, compression/ decompression process will continue trying without UADK.

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:53]: Added

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Fixed, thanks~

File: src/compressor/UadkAccel.h
Comment (Svelar) [reply src/compressor/UadkAccel.h]: Fixed

File: cmake/modules/Builduadk.cmake
Comment (Svelar) [reply cmake/modules/Builduadk.cmake:8]: UADK needs numactl, OpenSSL and other libraries, IMO dynamic library is more lighter-weight and easier.

Comment (cbodley) [reply cmake/modules/Builduadk.cmake]: static linkage for uadk doesn't mean it has to statically link openssl and its other deps, does it?

> that would avoid the extra effort to a) package and install the shared lib with ceph packages, and b) rename things to avoid conflicts with existing libraries installed on the system

shared libraries are more complicated for ceph to bundle. if you want to leave it this way, you'll need to rename `libwd.so`/`libwd_comp.so` to `libceph_wd.so`/`libceph_wd_comp.so` to avoid conflicts, and add them to `ceph.spec.in` and `debian/ceph-common.install` so they're installed along with other ceph binaries/libraries

but as we're currently only linking these into a single target (ceph_zlib), we won't actually benefit from a shared library

if/when this library is packaged by our supported distros, we can rely on the shared libraries they provide instead of bundling this way

File: CMakeLists.txt
Comment (cbodley) [reply CMakeLists.txt]: i hate to block progress here on cmake/packaging concerns. if you'd prefer to leave this OFF by default, we can follow up on that stuff in a separate pr

Comment (Svelar) [reply CMakeLists.txt]: https://github.com/ceph/ceph/blob/c4fb062db89e2598627cde62d455f397dc3377fc/CMakeLists.txt#L385  No objection at all. It has already ON by default, not sure why it is still showing OFF here, plz check the latest code.

File: cmake/modules/Builduadk.cmake
Comment (Svelar) [reply cmake/modules/Builduadk.cmake]: You're right, I forget such rpm matters. Committed a new version, please check~

**Comment by cbodley** [comment]
thanks, my comments were all addressed 👍 

i pushed this branch to ceph-ci so we can test the rpm/deb builds: https://shaman.ceph.com/builds/ceph/pr-58336/775ef028212a32d586e568a8af365d4284fa8c46/

File: src/compressor/zlib/ZlibCompressor.cc
Comment (cbodley) [reply src/compressor/zlib/ZlibCompressor.cc:211]: i think it's probably better to return the error directly so it's clear that something's broken. we don't want to get stuck in a situation where compression performs poorly because we're trying UADK and falling back every time

**Comment by cbodley** [comment]
> i pushed this branch to ceph-ci so we can test the rpm/deb builds: https://shaman.ceph.com/builds/ceph/pr-58336/775ef028212a32d586e568a8af365d4284fa8c46/

the builds are all green, but i don't see any evidence that UadkAccel.cc is being compiled. the 'make check' and 'ceph API tests' don't seem to be building it either

**Comment by Svelar** [comment]
> > i pushed this branch to ceph-ci so we can test the rpm/deb builds: https://shaman.ceph.com/builds/ceph/pr-58336/775ef028212a32d586e568a8af365d4284fa8c46/
> 
> the builds are all green, but i don't see any evidence that UadkAccel.cc is being compiled. the 'make check' and 'ceph API tests' don't seem to be building it either

Yes, only for aarch64, WITH_UADK is ON by default. Should we always keep it ON when running make check for any architectures?

File: src/compressor/zlib/ZlibCompressor.cc
Comment (Svelar) [reply src/compressor/zlib/ZlibCompressor.cc:211]: What do you think @idryomov ?

**Comment by cbodley** [comment]
> Yes, only for aarch64, WITH_UADK is ON by default

ok, i had misread that as 'default to ON _unless_ aarch64'. why not enable by default?

p.s. from the arm64 shaman build, i see the uadk build failing:
```
wd.c:18:10: fatal error: numa.h: No such file or directory
   18 | #include <numa.h>
      |          ^~~~~~~~
compilation terminated.
```

**Comment by Svelar** [comment]
> ok, i had misread that as 'default to ON _unless_ aarch64'. why not enable by default?

Till now, UADK is only provided by some specific ARM suppliers, e.g. Kunpeng platform. So I only enabled it on ARM.
> p.s. from the [arm64 shaman build](https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=arm64,AVAILABLE_ARCH=arm64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/81312//consoleFull), i see the uadk build failing:
> 
> ```
> wd.c:18:10: fatal error: numa.h: No such file or directory
>    18 | #include <numa.h>
>       |          ^~~~~~~~
> compilation terminated.
> ```

Seems like `numactl` should be dependency, will revise.

**Comment by cbodley** [comment]
> Till now, UADK is only provided by some specific ARM suppliers, e.g. Kunpeng platform. So I only enabled it on ARM.

i see, thanks. we don't really do any testing for arm, so maybe it's better to leave this off by default

**Comment by cbodley** [comment]
> I changed to install it no matter seastar's status, could you help trigger again?

pending in https://shaman.ceph.com/builds/ceph/pr-58336/b11309cf1fe1275fef93de1bb5c25cd47644569f/

File: src/compressor/zlib/ZlibCompressor.cc
Comment (idryomov) [reply src/compressor/zlib/ZlibCompressor.cc:211]: I agree with Casey that it's better to return the error, like it's done for other compressors.  That is why I commented in the first place ;)

File: src/compressor/UadkAccel.cc
Comment (idryomov) [reply src/compressor/UadkAccel.cc:373]: ... but you are not inheriting from `Compressor`.  `UadkAccel::decompress()` is an ordinary method that is called from `ZlibCompressor::decompress()` without any sort of virtual dispatch involved.  If the argument isn't passed, why pass it?

File: src/compressor/UadkAccel.cc
Comment (idryomov) [reply src/compressor/UadkAccel.cc:277]: > Yes, handle in `wd_comp_alloc_sess` of UADK.

I'm having trouble understanding the answer.  If `create_comp_session()` can fail, what is returned?  Can it be a NULL pointer?  Is it safe to pass what is returned to `uadk_do_compress()` and from there to `wd_do_comp_strm()`?

Comment (Svelar) [reply src/compressor/UadkAccel.cc:277]: `create_comp_session()` would return 0 if failed. 
`wd_do_comp_strm()` would call `wd_comp_check_params()` to check parameters, so it is safe.

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:373]: You are right. Sorry for missing it. I removed all unused parameters and functions.

File: src/compressor/zlib/ZlibCompressor.cc
Comment (Svelar) [reply src/compressor/zlib/ZlibCompressor.cc:211]: NP, revised.

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

File: CMakeLists.txt
Comment (Svelar) [reply CMakeLists.txt]: Hi @cbodley, for user-facing docs, we want to treat it carefully and make it as understandable as possible, so we will open another doc-related PR separately, and seek your advice then.

**Review by cbodley — APPROVED** [review]
cmake/packaging stuff looks great, thanks. we'll probably want to squash some of the commits before merge

**Comment by Svelar** [comment]
> cmake/packaging stuff looks great, thanks. we'll probably want to squash some of the commits before merge

Thanks for your approval. Once @idryomov approved, I will do the squash.

File: src/compressor/UadkAccel.cc
Comment (idryomov) [inline src/compressor/UadkAccel.cc:257]: bool -> int conversion is guaranteed to produce 1 for true and 0 for false:
```suggestion
  req.last = last_packet;
```

File: src/compressor/UadkAccel.cc
Comment (idryomov) [inline src/compressor/UadkAccel.cc:286]: No need to spell out true and false here:
```suggestion
    bool last_ptr = (i == in.buffers().end());
```

File: src/compressor/UadkAccel.cc
Comment (idryomov) [inline src/compressor/UadkAccel.cc:298]: Add a comment like in the other compressors -- it's kind of hard to guess what this "begin" business is about, especially since it messes up the alignment:
```suggestion
        // put a compressor variation mark in front of compressed stream, not used at the moment
        ptr.c_str()[0] = 0;
```

**Comment by idryomov** [comment]
> Thanks for your approval. Once @idryomov approved, I will do the squash.

@Svelar I left a few more nits, but note that this PR already has a

> No objection from RBD.

ack.  I'm not in the approval path here, you don't need to wait for me ;)

**Comment by Svelar** [comment]
> > Thanks for your approval. Once @idryomov approved, I will do the squash.
> 
> @Svelar I left a few more nits, but note that this PR already has a
> 
> > No objection from RBD.
> 
> ack. I'm not in the approval path here, you don't need to wait for me ;)

@idryomov Your review greatly improves security, readability and maintainability of the code, I'd like to address all your comments before merging it. Thanks for your time taking care of my code~

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:14]: legacy code, cleared.

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: reasonable, removed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: done

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Yes, modified.

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Keep same~

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: fixed

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: OK

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc]: Modified

File: src/compressor/UadkAccel.cc
Comment (Svelar) [reply src/compressor/UadkAccel.cc:275]: Added comments.

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test make check

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
@idryomov Not urging, but if you could review again, I'd be very grateful.

File: src/common/options/global.yaml.in
Comment (idryomov) [inline src/common/options/global.yaml.in:793]: I'd move this down so that it's placed after `qat_compressor_busy_polling` to avoid UADK-related options being defined in the middle of QAT-related options (namely between `qat_compressor_session_max_number` and `qat_compressor_busy_polling`).

**Review by idryomov — APPROVED** [review]
All of my comments seem to have been addressed, thanks for following up @Svelar!

**Comment by idryomov** [comment]
I'd squash most, if not all of the commits.  Since this is a pure addition of new code (with the exception of one `BuildRequires` line that got moved), a single commit probably makes the most sense.

In any case, please get rid of the merge commit -- we generally try to keep git history as linear as possible.

**Comment by idryomov** [comment]
@rzarzynski @ljflores This needs an ack from Core.

**Comment by rzarzynski** [comment]
@aclamk and @ifed01 will be interested as well. CCing.

File: src/common/options/global.yaml.in
Comment (Svelar) [reply src/common/options/global.yaml.in:798]: adjusted orders

**Comment by Svelar** [comment]
> I'd squash most, if not all of the commits. Since this is a pure addition of new code (with the exception of one `BuildRequires` line that got moved), a single commit probably makes the most sense.
> 
> In any case, please get rid of the merge commit -- we generally try to keep git history as linear as possible.

Of course, squashed.

**Comment by Svelar** [comment]
Build on: https://shaman.ceph.com/builds/ceph/wip-svelar-testing-1-2024-08-12-1856/

Hi @cbodley , because there is no ARM teuthology server available in the sepia lab, uadk option will not be enabled, no need for qa?

**Comment by Svelar** [comment]
jenkins test make check arm64

**Comment by Svelar** [comment]
jenkins test api

**Comment by Svelar** [comment]
Any more comments here? @rzarzynski @ljflores @ifed01 @aclamk

**Pull request merged**
